### ISSUE TITLE
04 05 06 Proyectos Diabetes

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/crerov-intro-machine-learning#4]
### **EDA + Decision Tree (cris04-eda-diabetes.ipynb)**

**Puntos fuertes:**
- EDA completo: histogramas, boxplots, heatmap de correlaciones
- Análisis de valores 0 (identificas correctamente que son datos faltantes)
- Comparas **gini, entropy, log_loss** (los 3 criterios)
- GridSearchCV con `max_depth`, `min_samples_split`, `min_samples_leaf`
- Pipeline con `SimpleImputer` para imputación de medianas
- Visualización del árbol con `plot_tree`
- Matriz de confusión con análisis interpretativo
- **Guardas el modelo** con pickle
- `stratify=y` en train_test_split

**Podrías:**
- Incluir`criterion` en el GridSearchCV (lo comparas manualmente por separado)
- Usar `class_weight` para el desbalanceo

**Nota honesta:** "Accuracy has gone down xd" - buena observación, el modelo base a veces supera al optimizado en datasets pequeños.

---

### **Random Forest (cris05-random-forest-diabetes.ipynb.ipynb)**

**Puntos fuertes:**
- GridSearchCV con **5 hiperparámetros** (n_estimators, max_depth, min_samples_split, min_samples_leaf, max_features)
- Visualización de primeros 4 árboles del bosque
- Feature importance con análisis interpretativo
- Matriz de confusión con interpretación médica
- **Guarda el modelo** con pickle
- Buena reflexión: "Medically, this is a problem because the model struggles to identify patients with diabetes"

**Podrías:**
- Usar `class_weight` para el desbalanceo
- Explorar `scoring='recall'` dado el contexto médico

---

### **Boosting (cris06-boosting-diabetes.ipynb)**

**Puntos fuertes:**
- Usas **HistGradientBoostingClassifier** (algoritmo moderno de sklearn)
- GridSearchCV con 4 hiperparámetros (max_iter, learning_rate, max_depth, max_features)
- **Guardas el modelo**

**Problemas:**
- **No hay comparación final de los 3 modelos** - requisito del proyecto
- Observación sin explorar: "Optimization makes the parameters go down. (??????????????)" - no investigas por qué, en ocasiones se da.
- Notebook muy corto, sin análisis de feature importance
- No pruebas otros algoritmos de boosting (XGBoost, GradientBoosting, AdaBoost)

---

### **Lo que destacas:**

1. **Interpretación médica** - Entiendes que el recall bajo es problemático en diagnóstico de diabetes
2. **Consistencia** - Guardas todos los modelos y datos procesados
3. **Visualización de RF** - Muestrsa los primeros 4 árboles del bosque (única en la cohorte)

---

### **Lo que falta:**

1. **Comparación final de los 3 modelos** - El enunciado del Boosting dice explícitamente:
   > "Paso 4: Analiza y compara resultados - Realiza un estudio comparativo de los tres modelos"

2. **Explorar el problema de optimización** - En Boosting observa que la optimización empeora los resultados pero no investigas por qué (overfitting del GridSearchCV)

---

### **Código que falta - Comparación final (en Boosting):**

```python
import pandas as pd

# Cargar los 3 modelos
import pickle
with open("/workspaces/.../models/decission_tree.pkl", 'rb') as f:
    model_dt = pickle.load(f)
with open("/workspaces/.../models/random_forest.pkl", 'rb') as f:
    model_rf = pickle.load(f)
with open("/workspaces/.../models/boosting.pkl", 'rb') as f:
    model_boost = pickle.load(f)

# Comparar resultados
models = {'Decision Tree': model_dt, 'Random Forest': model_rf, 'HistGradientBoosting': model_boost}
results = []

for name, model in models.items():
    y_pred = model.predict(X_test)
    results.append({
        'Model': name,
        'Accuracy': accuracy_score(y_test, y_pred),
        'Recall (Diabetes)': classification_report(y_test, y_pred, output_dict=True)['1']['recall'],
        'Precision (Diabetes)': classification_report(y_test, y_pred, output_dict=True)['1']['precision']
    })

comparison_df = pd.DataFrame(results)
print(comparison_df.to_string(index=False))

```
+ Conclusión en markdown

El Boosting necesita más desarrollo y la comparación final de los 3 modelos es un requisito faltante importante. El resto de cositas enhorabuena, son mínimas.  @crerov
